src/include/migraphx/op/scatter.hpp

```diff
@@ -33,29 +44,41 @@ struct scatter
         return {{"normalize_axes", normalize}};
     }
 
-    std::string name() const { return "scatter"; }
-
     shape normalize_compute_shape(std::vector<shape> inputs) const
     {
         check_shapes{inputs, *this}.has(3).standard();
-        return inputs.front();
+        // If non-packed, this converts to a packed output while preserving permutation of tensor
+        return inputs.front().with_lens(inputs.front().lens());
     }
 
     argument compute(const shape& output_shape, std::vector<argument> args) const
     {
         argument result{output_shape};
-        // max dimension in axis
+        auto& self = static_cast<const Derived&>(*this);
+
+        // max dimension in each axis
         auto axis_dim_size = output_shape.lens()[axis];
+        // cast all arguments as correct type
         visit_all(result, args[0], args[2])([&](auto output, auto data, auto update) {
+            // copy all of data to output
             std::copy(data.begin(), data.end(), output.begin());
             args[1].visit([&](auto indices) {
                 auto ind_s = indices.get_shape();
+                // iterate through items in index
                 shape_for_each(ind_s, [&](const auto& idx) {
-                    auto out_idx  = idx;
-                    auto index    = indices[ind_s.index(idx)];
+                    auto out_idx = idx;
+                    // Overloaded tensor_view::() invokes indexing logic of  std::size_t
+                    // shape::index(std::size_t i) const which handles nonstandard shapes correctly
+                    auto index = indices(idx.begin(), idx.end());
+
+                    // normalize negative indexes (may be redundant after using
+                    // normalize_compute_shape())
                     index         = (index < 0) ? index + axis_dim_size : index;
                     out_idx[axis] = index;
-                    output[output_shape.index(out_idx)] = update[ind_s.index(idx)];
+
+                    // look up the appropriate location in output, using index.
+                    // call reduction() method of derived struct to copy and reduce that element
+                    self.reduction()(output[output_shape.index(out_idx)], update[ind_s.index(idx)]);
```

Comment:
The access for `output` and `update` needs to be updated as well.

Reply:
Actually, I just pushed it to see if it would break any other tests.  But it's really not clear how to do the corresponding indexing into `output` and `update`.

Reply:
It would be `output(out_idx.begin(), out_idx.end())` and `update(idx.begin(), idx.end())`

Reply:
This is done and is in the process of passing tests, but 2 new questions:  

1. the iteration value `ind_s` seen in this snippet:

```
        visit_all(result, args[0], args[2])([&](auto output, auto data, auto update) {
            // copy all of data to output
            std::copy(data.begin(), data.end(), output.begin());
            args[1].visit([&](auto indices) {
                auto ind_s = indices.get_shape(); <===== HERE
```
is no longer used in a formula, only as a looping range.  Does the changed indexing call mean that we now have an extra, unneeded layer of nested looping?

2. Is my comment `                    // normalize negative indexes (may be redundant after using
                    // normalize_compute_shape())` in error?  It's not clear to me if the array modified here is the same one that was normalized with `normalize_compute_shape()`.